libraries/shared/watcher/storage_watcher.go

```diff
@@ -17,73 +17,103 @@
-	"strings"
+	"time"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/sirupsen/logrus"
 
+	"github.com/vulcanize/vulcanizedb/libraries/shared/fetcher"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/storage"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/transformer"
 	"github.com/vulcanize/vulcanizedb/pkg/datastore/postgres"
-	"github.com/vulcanize/vulcanizedb/pkg/fs"
 )
 
 type StorageWatcher struct {
-	db           *postgres.DB
-	tailer       fs.Tailer
-	Queue        storage.IStorageQueue
-	Transformers map[common.Address]transformer.StorageTransformer
+	db             *postgres.DB
+	StorageFetcher fetcher.IStorageFetcher
+	Queue          storage.IStorageQueue
+	Transformers   map[common.Address]transformer.StorageTransformer
 }
 
-func NewStorageWatcher(tailer fs.Tailer, db *postgres.DB) StorageWatcher {
+func NewStorageWatcher(fetcher fetcher.IStorageFetcher, db *postgres.DB) StorageWatcher {
 	transformers := make(map[common.Address]transformer.StorageTransformer)
 	queue := storage.NewStorageQueue(db)
 	return StorageWatcher{
-		db:           db,
-		tailer:       tailer,
-		Queue:        queue,
-		Transformers: transformers,
+		db:             db,
+		StorageFetcher: fetcher,
+		Queue:          queue,
+		Transformers:   transformers,
 	}
 }
 
-func (watcher StorageWatcher) AddTransformers(initializers []transformer.StorageTransformerInitializer) {
+func (storageWatcher StorageWatcher) AddTransformers(initializers []transformer.StorageTransformerInitializer) {
 	for _, initializer := range initializers {
-		storageTransformer := initializer(watcher.db)
-		watcher.Transformers[storageTransformer.ContractAddress()] = storageTransformer
+		storageTransformer := initializer(storageWatcher.db)
+		storageWatcher.Transformers[storageTransformer.ContractAddress()] = storageTransformer
 	}
 }
 
-func (watcher StorageWatcher) Execute() error {
-	t, tailErr := watcher.tailer.Tail()
-	if tailErr != nil {
-		return tailErr
+func (storageWatcher StorageWatcher) Execute(rows chan utils.StorageDiffRow, errs chan error, queueRecheckInterval time.Duration) {
+	ticker := time.NewTicker(queueRecheckInterval)
+	go storageWatcher.StorageFetcher.FetchStorageDiffs(rows, errs)
+	for {
+		select {
+		case fetchErr := <-errs:
+			logrus.Warn(fmt.Sprintf("error fetching storage diffs: %s", fetchErr))
+		case row := <-rows:
+			storageWatcher.processRow(row)
+		case <-ticker.C:
+			storageWatcher.processQueue()
+		}
+	}
+}
+
+func (storageWatcher StorageWatcher) processRow(row utils.StorageDiffRow) {
+	storageTransformer, ok := storageWatcher.Transformers[row.Contract]
+	if !ok {
+		// ignore rows from unwatched contracts
+		return
 	}
-	for line := range t.Lines {
-		row, parseErr := utils.FromStrings(strings.Split(line.Text, ","))
-		if parseErr != nil {
-			return parseErr
+	executeErr := storageTransformer.Execute(row)
+	if executeErr != nil {
+		if isKeyNotFound(executeErr) {
```

Comment:
I'm not very experienced with the storage transformers, what is it that can change to break this error from continuing in `processQueue`?

Reply:
Not sure I understand the question - could you say more? `processQueue` differs from `processRow` in that it only takes further action if `exectureErr` _is_ nil - otherwise it just leaves the row in the queue to be retried later.

Reply:
Sorry that was worded poorly, I guess I am wondering what it is that changes in the storage transformer's execution cycle that causes the `executeErr` it returns to be nil in the `processQueue` when it had been `isKeyNotFound` here.

Reply:
Oh yeah great question! That's basically the crux of what makes this feature worthwhile :)

Storage keys for values in more complex data structures like mappings are derived by hashing the index of that variable on the contract concatenated with the key to the mapping. E.g. if the contract has a variable `mapping (address => uint256) balances`, you need to know the `address` keys in that mapping to be able to recognize a given diff as representing a changed balance.

We're using the event transformers to catch all of the possible keys being used in the mappings we care about. But the event transformers could be lagging behind the storage transformers, in which case we wouldn't yet be able to recognize a given diff's storage key. That's where we queue the diff, and then this process will keep retrying - on the assumption that the event transformers will eventually give us the value we need to recognize the key.

Reply:
Thank you, that clears everything up!